pkg/logs/sidecar/proxy.go

```diff
@@ -0,0 +1,250 @@
+
+	logs := make(chan events.LogChunk, logsBuffer)
+
+	// create stream for incoming logs
+
+	err := p.logsStream.Init(ctx)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		p.log.Debugw("logs proxy stream started")
+		err := p.streamLogs(ctx, logs)
+		if err != nil {
+			p.handleError(err, "proxy stream logs error")
+		}
+	}()
+
+	for l := range logs {
+		select {
+		case <-sigs:
+			p.log.Warn("logs proxy received signal, exiting", "signal", sigs)
+			return ErrStopSignalReceived
+		case <-ctx.Done():
+			p.log.Warn("logs proxy context cancelled, exiting")
+			return nil
+		default:
+			err = p.logsStream.Push(ctx, l)
+			if err != nil {
+				p.handleError(err, "error pushing logs to stream")
+				return err
+			}
+		}
+	}
+
+	p.log.Infow("logs proxy sending completed")
+	return nil
+}
+
+func (p *Proxy) streamLogs(ctx context.Context, logs chan events.LogChunk) (err error) {
+	pods, err := executor.GetJobPods(ctx, p.podsClient, p.executionId, 1, 10)
+	if err != nil {
+		p.handleError(err, "error getting job pods")
+		return err
+	}
+
+	for _, pod := range pods.Items {
+		l := p.log.With("namespace", pod.Namespace, "podName", pod.Name, "podStatus", pod.Status)
+
+		switch pod.Status.Phase {
+
+		case corev1.PodRunning:
+			l.Debug("streaming pod logs: immediately")
+			return p.streamLogsFromPod(pod, logs)
+
+		case corev1.PodFailed:
+			err := fmt.Errorf("can't get pod logs, pod failed: %s/%s", pod.Namespace, pod.Name)
+			p.handleError(err, "streaming pod logs: pod failed")
+			return err
+
+		default:
+			l.Debugw("streaming pod logs: waiting for pod to be ready")
+			testFunc := p.isPodLoggable(pod.Name)
+			if err = wait.PollUntilContextTimeout(ctx, pollInterval, podStartTimeout, true, testFunc); err != nil {
+				// try to get pod container statuses from Waiting and Terminated states
+				status := p.getPodContainerStatuses(pod)
+				p.handleError(err, "can't get pod container status after pod failure")
+				return errors.Wrap(err, status)
+			}
+
+			l.Debug("streaming pod logs: pod is loggable")
+			return p.streamLogsFromPod(pod, logs)
+		}
+	}
+	return
+}
+
+func (p *Proxy) streamLogsFromPod(pod corev1.Pod, logs chan events.LogChunk) (err error) {
+	defer close(logs)
+
+	var containers []string
+	for _, container := range pod.Spec.InitContainers {
+		containers = append(containers, container.Name)
+	}
+
+	for _, container := range pod.Spec.Containers {
+		containers = append(containers, container.Name)
+	}
+
+	for _, container := range containers {
+
+		req := p.podsClient.GetLogs(
+			pod.Name,
+			&corev1.PodLogOptions{
+				Follow:     true,
+				Timestamps: true,
+				Container:  container,
+			})
+
+		stream, err := req.Stream(context.Background())
+		if err != nil {
+			p.handleError(err, "error getting pod logs stream")
+			return err
+		}
+
+		reader := bufio.NewReader(stream)
+		for {
+			b, err := utils.ReadLongLine(reader)
+			if err != nil {
+				if err == io.EOF {
```

Comment:
A safer check is `errors.Is`, but if this error is not wrapped anywhere `==` will also work

Reply:
done